content/algorithms/planarity/planarity.md

```diff
@@ -0,0 +1,939 @@
+
+Other equivalent representations for cellular embeddings include the ribbon graph, a topological space formed by gluing together topological disks for the vertices and edges of an embedded graph, and the graph-encoded map, an edge-colored cubic graph with four vertices for each edge of the embedded graph.
+
+## Neighbor ordering:
+In comparison to a usual graph structure, the embedding also stores the order of all neighbors for every vertex. The order of the neighbors can be given in clockwise (cw) direction or counterclockwise (ccw) direction. This order is stored as edge attributes in the underlying directed graph. For the edge (u, v) the edge attribute ‘cw’ is set to the neighbor of u that follows immediately after v in clockwise direction.
+
+In order for a PlanarEmbedding to be valid it must fulfill multiple conditions. It is possible to check if these conditions are fulfilled with the method check_structure(). The conditions are:
+1. Edges must go in both directions (because the edge attributes differ)
+2. Every edge must have a ‘cw’ and ‘ccw’ attribute which corresponds to a correct planar embedding.
+
+As long as a PlanarEmbedding is invalid only the following methods should be called:
+
+1. add_half_edge()
+2. connect_components()
+
+Even though the graph is a subclass of nx.DiGraph, it can still be used for algorithms that require undirected graphs, because the method is_directed() is overridden. This is possible, because a valid PlanarGraph must have edges in both directions.
+
+## Half edges:
+In methods like add_half_edge the term “half-edge” is used, which is a term that is used in doubly connected edge lists. It is used to emphasize that the edge is only in one direction and there exists another half-edge in the opposite direction. While conventional edges always have two faces (including outer face) next to them, it is possible to assign each half-edge exactly one face. For a half-edge (u, v) that is oriented such that u is below v then the face that belongs to (u, v) is to the right of this half-edge.
+
+---
+
+
+
+## Parameters:
+
+
+*   *incoming_graph_data :*  
+graph (optional, default: None)
+Data used to initialize the graph. If None (default), an empty graph is created. It can be an edge list or any NetworkX graph object. If the relevant Python packages are installed, it can also be a 2D NumPy array, a SciPy sparse matrix, or a PyGraphviz graph.
+*   *attr :*  
+keyword arguments, optional (default: no attributes)
+Attributes to add to the graph, provided as key-value pairs.
+
+---
+
+
+
+
+```python
+class PlanarEmbedding(nx.DiGraph):
```

Comment:
Including this code makes the notebook hard to read. It is code in NetworkX, so you can call it from that.  And it would be good to explain it rather than making the reader figure out how the code works.